perf/pkeyread.c

```diff
@@ -0,0 +1,381 @@
+                }
+            }
+        } else {
+            /* run all samples for one format (format_id) */
+            for (k = 0; k < SAMPLE_ALL; k++) {
+                sample_id = k;
+                if (!perflib_run_multi_thread_test(do_f[format_id], threadcount, &duration)) {
+                    fprintf(stderr, "Failed to run the test %s in %s format]\n",
+                        sample_names[k], format_names[format_id]);
+                    return EXIT_FAILURE;
+                }
+
+                if (err) {
+                    fprintf(stderr, "Error during test of %s in %s format\n",
+                        sample_names[k], format_names[format_id]);
+                    return EXIT_FAILURE;
+                }
+
+                if (verbose)
+                    printf("Average time per %s(%s) calls: %lfus\n",
+                        fname[format_id], sample_names[k], get_avcalltime());
+                else
+                    printf("[%s %s] %lf\n", sample_names[k],
+                        format_names[format_id], get_avcalltime());
+            }
+        }
+    } else if (format_id == FORMAT_ALL) {
+        /* run one sample for all formats */
+        for (f = 0; f < FORMAT_ALL; f++) {
+            sample_id = k;
+            if (!perflib_run_multi_thread_test(do_f[f], threadcount, &duration)) {
+                fprintf(stderr, "Failed to run the test %s in %s format]\n",
+                    sample_names[key_id], format_names[f]);
+                return EXIT_FAILURE;
+            }
+
+            if (err) {
+                fprintf(stderr, "Error during test of %s in %s format\n",
+                    sample_names[key_id], format_names[f]);
+                return EXIT_FAILURE;
+            }
+
+            if (verbose)
+                printf("Average time per %s(%s) calls: %lfus\n",
+                    fname[f], sample_names[key_id], get_avcalltime());
+            else
+                printf("[%s %s] %lf\n", sample_names[key_id],
+                    format_names[f], get_avcalltime());
+        }
+    } else {
+        /* run one sample in desired format */
+        sample_id = key_id;
+        if (!perflib_run_multi_thread_test(do_f[format_id], threadcount, &duration)) {
+            fprintf(stderr, "Failed to run the test %s in %s format]\n",
+                sample_names[key_id], format_names[format_id]);
+            return EXIT_FAILURE;
+        }
+
+        if (err) {
+            fprintf(stderr, "Error during test of %s in %s format\n",
+                sample_names[key_id], format_names[format_id]);
+            return EXIT_FAILURE;
+        }
+        if (verbose)
+            printf("Average time per %s(%s) calls: %lfus\n",
+                fname[format_id], sample_names[key_id], get_avcalltime());
+        else
+            printf("[%s %s] %lf\n", sample_names[key_id],
+                format_names[format_id], get_avcalltime());
+    }
```

Comment:
I disklike all the code duplication here. We have four variants of essentially the same code. Surely we can collapse all this into one variant?

Reply:
The duplicate code which prints result got moved to `report_result() function. Not sure what you mean here. There are four different cases we want to test:
  - single key type, single format
  - single key type, both formats
  - all key types, single format
  - all key types all formats

I don't know how to combine those four cases to limit duplication.